rubix-spi/src/main/java/com/qubole/rubix/spi/fop/ObjectPoolPartition.java

```diff
@@ -58,40 +61,123 @@ public ObjectPoolPartition(ObjectPool<T> pool, PoolConfig config,
+          throw new RuntimeException("Cannot get a free object from the pool");
+        }
+      }
+    }
+    catch (InterruptedException e) {
+      throw new RuntimeException(e); // will never happen
+    }
+
+    freeObject.setLastAccessTs(System.currentTimeMillis());
+    return freeObject;
+  }
+
+  private synchronized Poolable<T> increaseObjects(int delta, boolean returnObject)
   {
     int oldCount = totalCount;
     if (delta + totalCount > config.getMaxSize()) {
       delta = config.getMaxSize() - totalCount;
     }
+
+    Poolable<T> objectToReturn = null;
     try {
       for (int i = 0; i < delta; i++) {
         T object = objectFactory.create(host, socketTimeout, connectTimeout);
         if (object != null) {
-          objectQueue.put(new Poolable<>(object, pool, host));
+          // Do not put the first object on queue
+          // it will be returned to the caller to ensure it's request is satisfied first if object is requested
+          Poolable<T> poolable = new Poolable<>(object, pool, host);
+          if (objectToReturn == null && returnObject) {
+            objectToReturn = poolable;
+          }
+          else {
+            objectQueue.put(poolable);
+          }
           totalCount++;
         }
       }
-      log.debug("Increased pool size by " + (totalCount - oldCount) + " to new size: " + totalCount + ", current queue size: " + objectQueue.size());
+
+      if (delta > 0 && (totalCount - oldCount) == 0) {
+        log.warn(String.format("Could not increase pool size. Pool state: totalCount=%d queueSize=%d delta=%d", totalCount, objectQueue.size(), delta));
+      }
+      else {
+        log.debug(String.format("Increased pool size by %d, to new size: %d, current queue size: %d, delta: %d",
+                totalCount - oldCount, totalCount, objectQueue.size(), delta));
+      }
     }
-    catch (InterruptedException e) {
+    catch (Exception e) {
+      log.warn(String.format("Unable to increase pool size. Pool state: totalCount=%d queueSize=%d delta=%d", totalCount, objectQueue.size(), delta), e);
+      // objectToReturn is not on the queue hence untracked, clean it up before forwarding exception
+      if (objectToReturn != null) {
+        objectFactory.destroy(objectToReturn.getObject());
+        objectToReturn.destroy();
+      }
       throw new RuntimeException(e);
     }
-    return totalCount;
+
+    return objectToReturn;
   }
 
-  public synchronized boolean decreaseObject(Poolable<T> obj)
+  public boolean decreaseObject(Poolable<T> obj)
   {
+    checkState(obj.getHost() != null, "Invalid object");
+    checkState(obj.getHost().equals(this.host),
+            "Call to free object of wrong partition, current partition=%s requested partition = %s",
+            this.host, obj.getHost());
+    objectRemoved(obj);
     objectFactory.destroy(obj.getObject());
-    totalCount--;
+    obj.destroy();
     return true;
   }
 
+  private synchronized void objectRemoved(Poolable<T> obj)
```

Comment:
we can drop 'obj' now as it is not getting used. Also, may rename it to 'decreasseTotoalCount'.

Reply:
Kept it as an extension point, if we want to do anything specific to obj in future. Will remove it for now and add when needed

Reply:
will retain the same name as we are informing PoolPartition via this method that an object is removed. It can do more than decreasing totalCount, just that right it is only doing that.